### DISCUSSION TITLE
Manually define dependencies of a target II

### DISCUSSION THREAD

**Discussion by MxNl** [ropensci/targets#830] [Help] [answered]
This is a slight variation of the discussion [here](https://github.com/ropensci/targets/discussions/828).
I have a pipeline where one target (summary) uses two upstream targets as input. But I only want to rerun the target if the target rerun_summary has changed to run it every month (or minute as inthe example). But if the target is run, it should also use the up-to-date data. Or in other words: **run the target summary only if rerun_summary changes and not if data changes. But if it is rerun, use the up-to-date data**. Maybe I am not seeing the most obvious or simple solution :-). I thought of using something like tar_change() and use the target rerun_summary for the change argument, but this function has no pattern argument.

Example:
``` r
library(targets)
tar_script({
    library(lubridate)
    library(dplyr)
    make_summary <- function(data, deps) {
        length(deps)
        summarise(data, mean = mean(x, na.rm = TRUE))
    }
    list(
        tar_target(
            data, 
            tar_group(group_by(data.frame(x = 11:20, groups = rep_len(c("a", "b"), length.out = 10)), groups)), 
            iteration = "group"), 
        tar_target(
            rerun_summary, 
            lubridate::minute(Sys.time()), 
            cue = tar_cue("always")), 
        tar_target(
            summary, 
            make_summary(data, rerun_summary), 
            pattern = data, 
            iteration = "group")
    )
})
tar_make()
#> 
#> Attache Paket: 'lubridate'
#> 
#> Die folgenden Objekte sind maskiert von 'package:base':
#> 
#>     date, intersect, setdiff, union
#> 
#> 
#> Attache Paket: 'dplyr'
#> 
#> Die folgenden Objekte sind maskiert von 'package:stats':
#> 
#>     filter, lag
#> 
#> Die folgenden Objekte sind maskiert von 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> * start target data
#> * built target data
#> * start target rerun_summary
#> * built target rerun_summary
#> * start branch summary_54d58009
#> * built branch summary_54d58009
#> * start branch summary_f14045dd
#> * built branch summary_f14045dd
#> * built pattern summary
#> * end pipeline
```

<sup>Created on 2022-04-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by wlandau** [comment]
It's a bit of a hack, but you can use `data_constant` as a dependency and actually read `data` instead. Sketch:

```r
# _targets.R file
library(targets)
list(
  tar_target(data, ...),
  tar_target(data_constant, data, cue = tar_cue(mode = "never")),
  tar_target(rerun_summary, ...),
  tar_target(
    summary, {
      data_constant # dependency of summary
      rerun_summary
      mean(eval(parse(text = "targets::tar_read(data)"))$x, na.rm = TRUE) # "data" is part of the text, no longer a symbol in the code
    },
    pattern = data_constant
  )
)
```

**Reply by MxNl** [reply]
You are a genius. Is there anything you can't figure out ? :-) Thanks

**Comment by uhkeller** [comment]
Sorry for resurrecting this old topic. I'm looking for a way to build a dynamic branched target if and only if one specific (simple, non-branching) dependency changes, but *not* when only the grouped dependency changes. This seemed to be the same as the original question here. But I'm struggling to understand how (if?) the solution works. When I combine the original problem with the "sketch", I get this:

```R
# _targets.R file
library(targets)
library(lubridate)
library(dplyr)
list(
  tar_target(
    data,
    tar_group(group_by(data.frame(x = 11:20, groups = rep_len(c("a", "b"), length.out = 10)), groups)),
    iteration = "group"
  ),
  tar_target(data_constant, data, cue = tar_cue(mode = "never")),
  tar_target(rerun_summary, lubridate::minute(Sys.time()), 
    cue = tar_cue("always")
  ),
  tar_target(
    summary,
    {
      data_constant # dependency of summary
      rerun_summary
      mean(eval(parse(text = "targets::tar_read(data)"))$x, na.rm = TRUE) # "data" is part of the text, no longer a symbol in the code
    },
    pattern = data_constant
  )
)
```

`summary` is the target that I want to make if and only if dependency `rerun_summary` changes, but not when only the grouped dependency `data` changes.

When I run this, `summary` has length 10, corresponding to the number of lines in `data`, but it should have length 2, corresponding to the number of groups in `data`.  If I change `pattern = data_constant` to `pattern = data`, `summary` is computed correctly, but of course now it gets recomputed every time `data` changes.

I tried other ways of achieving the same goal. `tar_change()` looked like a good candidate, but it does not have a `pattern` argument so it appears it cannot handle dynamic branching. `tar_cue_skip()` also seemed promising, but the only way to check if `rerun_summary` has changed seems to be `tar_outdated()` and that can't be used inside `_targets.R` scripts.

To give simpler example, in the following script I want rebuild `means` if and only if `run_means` has changed. How can I do this?

```R
# _targets.R file
library(targets)
tar_option_set(seed = NA)
list(
  tar_target(
    data,    
    replicate(3, rnorm(2), simplify = FALSE)
  ),
  tar_target(run_means, 1),
  tar_target(means, mean(data[[1]]), pattern = map(data))
)
```

**Reply by wlandau** [reply]
> When I run this, summary has length 10, corresponding to the number of lines in data, but it should have length 2, corresponding to the number of groups in data. If I change pattern = data_constant to pattern = data, summary is computed correctly, but of course now it gets recomputed every time data changes.

You may have better luck if you set `iteration = "group"` in the `data_constant` target. I suspect that and https://github.com/ropensci/targets/discussions/830#discussioncomment-2604424 will solve your instance of the issue.

**Reply by uhkeller** [reply]
Thanks. However, with this change it still does not work, though it fails in a different way. It becomes more obvious if I change the grouping by ordering the grouping variable in `data`:

```R
library(targets)
library(lubridate)
library(dplyr)
list(
  tar_target(
    data,
    tar_group(
      # EDIT to order grouping variable:
      group_by(
        data.frame(x = 11:20, groups = sort(rep_len(c("a", "b"), length.out = 10))),
        groups
      )
    ),
    iteration = "group"
  ),
  tar_target(data_constant, data, cue = tar_cue(mode = "never"), iteration = "group"),
  # EDIT to force rebuilding of `summary` without having to wait:
  tar_target(rerun_summary, 1, cue = tar_cue("always")),
  tar_target(
    summary,
    {
      data_constant # dependency of summary
      rerun_summary
      mean(eval(parse(text = "targets::tar_read(data)"))$x, na.rm = TRUE) # "data" is part of the text, no longer a symbol in the code
    },
    pattern = map(data_constant)
  )
)
```

```
r$> tar_destroy()
[…]
r$> tar_make()
[…]
r$> tar_read(summary)
summary_3e478f75 summary_02b68572 
            15.5             15.5 
r$> tar_read(data)
# A tibble: 10 × 3
       x groups tar_group
   <int> <chr>      <int>
 1    11 a              1
 2    12 a              1
 3    13 a              1
 4    14 a              1
 5    15 a              1
 6    16 b              2
 7    17 b              2
 8    18 b              2
 9    19 b              2
10    20 b              2
```

The means in `summary` are calculated over all elements of `x` instead of by group. They should be 13 and 18, not 15.5.

As you pointed out, even if this worked (and I'm sure it can be made to work somehow), it would be "a bit of a hack" and not especially readable. I hope you will consider adding an "official" way to manually define dependencies of dynamic targets, maybe by adding the `pattern` argument to `tar_change()`.

**Reply by uhkeller** [reply]
Found a way to make it work: while `tar_change()` doesn't have a `pattern` argument, it *does* have an `iteration` argument. So this works:

```R
library(targets)
library(tarchetypes)
library(dplyr)
list(
  tar_target(
    data,
    tar_group(
      group_by(
        data.frame(x = 11:20, groups = sort(rep_len(c("a", "b"), length.out = 10))),
        groups
      )
    ),
    iteration = "group"
  ),
  tar_target(rerun_summary, 1),
  tar_change(
    data_constant,
    command = eval(parse(text = "tar_read(data)")),
    change = rerun_summary,
    iteration = "group"
  ),
  tar_target(
    summary,
    mean(data_constant$x),
    pattern = map(data_constant)
  )
)
```

**Reply by wlandau** [reply]
Glad that worked.

A comment on this whole thread: if `tar_read()` works without obvious errors for you, then that's fine. But its may not always work for all pipelines. I am working on cloud computing support for `targets`, and I am not sure if it will work there. As I mentioned in https://github.com/ropensci/targets/discussions/830#discussioncomment-2604424, `eval(parse(text = "tar_read(data)"))` is a hack because `tar_load()` and `tar_read()` are not really meant to be called from inside a target.

**Reply by uhkeller** [reply]
Thanks for the heads up! All the more reason to implement a way to properly define manual dependencies, don't you think? 😉

**Reply by wlandau** [reply]
It is extremely rare to need to update on one upstream target but not the other, and I have no plans to implement in `targets`.

**Comment by wlandau** [comment, marked as answer]
Looking back at the original question, the request has a tough set of uncommon competing requirements:

1. Update on changes to `rerun_summary`.
2. Dynamically branch over `data`.
3. Don't update on changes to `data`.
4. I would add: avoid `tar_read()` and `tar_load()`.

Let's drop (2) to simplify things. The pipeline becomes:

```r
# _targets.R file
library(targets)
tar_option_set(packages = c("dplyr", "lubridate"))
make_summary <- function(data, deps) {
  length(deps)
  summarise(data, mean = mean(x, na.rm = TRUE))
}
list(
  tar_target(
    data, 
    data.frame(x = 1)
  ),
  tar_target(
    rerun_summary, 
    lubridate::minute(Sys.time()), 
    cue = tar_cue("always")
  ), 
  tar_target(
    summary, 
    make_summary(data, rerun_summary)
  )
)
```

The workaround I can recommend for this rare case is to write the data to a file without tracking it (i.e. avoid `format = "file"`).

```r
# _targets.R file:
library(targets)
tar_option_set(packages = c("dplyr", "lubridate"))
make_summary <- function(data, deps) {
  length(deps)
  summarise(data, mean = mean(x, na.rm = TRUE))
}
list(
  tar_target(
    data, 
    saveRDS(data.frame(x = 1), "data.rds")
  ),
  tar_target(
    rerun_summary, 
    lubridate::minute(Sys.time()), 
    cue = tar_cue("always")
  ), 
  tar_target(
    summary, 
    make_summary(readRDS("data.rds"), rerun_summary)
  )
)
```

Generalizing the file-based `data` target to dynamic branching is unfortunately painful, but it is doable, and the need rarely arises.

Going forward, because of long-term plans for cloud integration, I plan to only allow `tar_load()` and `tar_read()` from inside targets when those targets have `repository = "local"` and `format = "file"` (or `"file_fast"`).
